index.js

```diff
@@ -29,75 +29,43 @@ function fastifyCors (fastify, opts, next) {
   }, opts)
 
-  const isOriginFalsy = !origin
-  const isOriginString = typeof origin === 'string'
-  const isOriginFunction = typeof origin === 'function'
+  const resolveOriginOption = typeof origin === 'function' ? resolveOriginWrapper : (_, cb) => cb(null, origin)
 
-  if (preflight === true) {
-    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
-      // Do not handle preflight requests if the origin was not allowed
-      if (!req.corsOriginAllowed) {
-        reply.code(404).type('text/plain').send('Not Found')
-        return
-      }
-
-      // Strict mode enforces the required headers for preflight
-      if (strictPreflight === true && (!req.headers.origin || !req.headers['access-control-request-method'])) {
-        reply.status(400).type('text/plain').send('Invalid Preflight Request')
-        return
-      }
-
-      // Handle preflight headers
-      reply.header(
-        'Access-Control-Allow-Methods',
-        Array.isArray(methods) ? methods.join(', ') : methods
-      )
-
-      if (allowedHeaders === null) {
-        vary(reply, 'Access-Control-Request-Headers')
-        const reqAllowedHeaders = req.headers['access-control-request-headers']
-        if (reqAllowedHeaders !== undefined) {
-          reply.header('Access-Control-Allow-Headers', reqAllowedHeaders)
-        }
-      } else {
-        reply.header(
-          'Access-Control-Allow-Headers',
-          Array.isArray(allowedHeaders) ? allowedHeaders.join(', ') : allowedHeaders
-        )
-      }
-
-      if (maxAge !== null) {
-        reply.header('Access-Control-Max-Age', String(maxAge))
-      }
+  fastify.decorateRequest('corsOriginAllowed', undefined)
+  fastify.addHook('onRequest', onRequest)
 
-      // Safari (and potentially other browsers) need content-length 0,
-      // for 204 or they just hang waiting for a body
-      reply
-        .code(optionsSuccessStatus)
-        .header('Content-Length', '0')
-        .send()
-    })
+  if (preflight === true) {
+    fastify.options('*', { schema: { hide: hideOptionsRoute } }, preflightHandler)
   }
 
-  fastify.decorateRequest('corsOriginAllowed', undefined)
+  next()
 
-  fastify.addHook('onRequest', onRequest)
   function onRequest (req, reply, next) {
     // Always set Vary header
     // https://github.com/rs/cors/issues/10
     vary(reply, 'Origin')
 
-    if (isOriginFalsy) {
-      req.corsOriginAllowed = false
-      return next()
-    }
+    resolveOriginOption(req, (error, resolvedOriginOption) => {
+      if (error !== null) {
+        return next(error)
+      }
+
+      // Disable CORS and preflight if false
+      if (resolvedOriginOption === false) {
+        req.corsOriginAllowed = false
+        return next()
+      }
 
-    configureOrigin(req, reply, (err, origin) => {
-      if (err !== null) return next(err)
+      // Falsy values are invalid
+      if (!resolvedOriginOption) {
+        return next(new Error('Invalid CORS origin option'))
```

Comment:
This is new behavior? In past code, we are setting `req.corsOriginAllowed = false` on falsy values.

Reply:
If is an intentional behavior, should be tested as well.

Reply:
You are correct that this is a functional change. Before, if the origin option was a falsy value, CORS would just be disabled. If the origin option was a function that returned a falsy value, the accept origin header would be `*`.

This was an attempt to make the behavior from the origin option predictable in the following ways:
1. The same behavior if the origin option is a value OR a function 
2. If the option is an invalid value (defined as any value not defined in the documentation), an error is thrown. 

Clearly there is some risk to a change like this. However I would argue that it will actual make the module more understandable because the behavior is consistent. I welcome any discussion on this point 😄

Reply:
If your goal is to "refactor for clarity" then you should not also be introducing new functionality. Doing so makes the PR way more difficult to evaluate.

Reply:
I agree with @jsumners. I would like to see this change, but it's better in another PR.

Reply:
I agree with your point.

The challenge is that the old behavior was a bit tricky (and possibly unintentional):
* if the `origin` option is falsy, CORS is disabled :ballot_box_with_check: 
* if the `origin` option is a function that returns `false`, CORS is disabled :ballot_box_with_check: 
* if the `origin` option is a function that returns a falsy value **other than** `false`, the allowed origin is `*`. (Note this behavior was never documented or tested) :scream: 

:point_up: Handling the last case will add a lot of noise, but should be possible... I'll see if I can separate it into 2 PRs...

Reply:
It may not be ideal to replicate weird behavior. But consider that you have reviewers who probably know less about the code than you do at this point and they need to read the original code to understand the new code in order to properly review it. Any functionality change will be a radical change in this respect, thus making it prohibitively more difficult to review. For this case, I would note the change I want to make and then submit a follow up PR after the refactor PR is accepted.

Reply:
@RafaelGSS @jsumners I removed this change. I'll put it into another PR.